lib/Doctrine/DBAL/Query/QueryParts.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\DBAL\Query;
+
+use Doctrine\DBAL\Query\Expression\CompositeExpression;
+
+/**
+ * @internal
+ */
+class QueryParts
+{
+    /** @var string[] */
+    public $select = [];
+
+    /** @var bool */
+    public $distinct = false;
+
+    /** @var From[] */
+    public $from = [];
+
+    /** @var string|null */
+    public $table;
+
+    /**
+     * Lists of joins indexed by from alias.
+     *
+     * @var array<string, Join[]>
+     */
+    public $join = [];
+
+    /** @var string[] */
+    public $set = [];
+
+    /** @var CompositeExpression|null */
+    public $where = null;
+
+    /** @var string[] */
+    public $groupBy = [];
+
+    /** @var CompositeExpression|null */
+    public $having = null;
+
+    /** @var string[] */
+    public $orderBy = [];
+
+    /** @var array<string, mixed> */
+    public $values = [];
+
+    /**
+     * Deep clone of all expression objects in the SQL parts.
+     */
+    public function __clone()
+    {
+        foreach ($this->from as $key => $from) {
+            $this->from[$key] = clone $from;
+        }
+
+        foreach ($this->join as $fromAlias => $joins) {
+            foreach ($joins as $key => $join) {
+                $this->join[$fromAlias][$key] = clone $join;
+            }
+        }
+
+        if ($this->where !== null) {
+            $this->where = clone $this->where;
+        }
+
+        if ($this->having !== null) {
+            $this->having = clone $this->having;
+        }
```

Comment:
**What about `$values`, should they be (deep-)cloned?**

The previous implementation blindly cloned objects and 1-level deep arrays of objects, so it looks like this also applied to `$sqlParts['values']`; was this intentional? I'm not sure.

Reply:
Before figuring this out, let's see why exactly we need a separate class. The builder object always uses the same single instance of parts.

Reply:
Do you want to merge QueryParts into QueryBuilder?

Reply:
Yes.

Reply:
Done in 42c60625a914dc223d27047afe10b2ea2df793ca.